drag_and_drop_v2/public/js/drag_and_drop.js

```diff
@@ -657,43 +641,48 @@ function DragAndDropBlock(runtime, element, configuration) {
 
     var placeItem = function($zone, $item) {
         var item_id;
-        var $anchor;
         if ($item !== undefined) {
             item_id = $item.data('value');
-            // Element was placed using the mouse,
-            // so use relevant properties of *item* when calculating new position below.
-            $anchor = $item;
         } else {
             item_id = $selectedItem.data('value');
-            // Element was placed using the keyboard,
-            // so use relevant properties of *zone* when calculating new position below.
-            $anchor = $zone;
         }
+
         var zone = String($zone.data('uid'));
         var zone_id = $zone.data('zone_id');
         var zone_align = $zone.data('zone_align');
-        var $target_img = $root.find('.target-img');
 
-        // Calculate the position of the item to place relative to the image.
-        var x_pos = $anchor.offset().left + ($anchor.outerWidth()/2) - $target_img.offset().left;
-        var y_pos = $anchor.offset().top + ($anchor.outerHeight()/2) - $target_img.offset().top;
-        var x_pos_percent = x_pos / $target_img.width() * 100;
-        var y_pos_percent = y_pos / $target_img.height() * 100;
+        var items_in_zone_count = countItemsInZone(zone);
+        if (configuration.max_items_per_zone && configuration.max_items_per_zone <= items_in_zone_count) {
+            state.last_action_correct = false;
+            state.feedback = gettext("This zone already contains maximum allowed number of items or more");
+            applyState();
+            return;
+        }
 
         state.items[item_id] = {
             zone: zone,
             zone_align: zone_align,
-            x_percent: x_pos_percent,
-            y_percent: y_pos_percent,
             submitting_location: true,
         };
         // Wrap in setTimeout to let the droppable event finish.
         setTimeout(function() {
             applyState();
-            submitLocation(item_id, zone, x_pos_percent, y_pos_percent);
+            submitLocation(item_id, zone);
         }, 0);
     };
 
+    var countItemsInZone = function(zone) {
+        var count = 0;
+        for (var key in state.items) {
+            if (state.items.hasOwnProperty(key)) {
```

Comment:
@itsjeyd `state.items` is an object, and iterating over object this way yields all "own" properties and all "inherited" properties (properties from prototype), mostly functions, liek `toString` and so on.

I believe the approach we used here was to use `Object.keys(object).forEach(...)` - I just have missed that and written it old-style :)